src/locked_mem.rs

```diff
@@ -37,51 +37,141 @@ impl<const N: usize> LockedArray<N> {
             z
         };
 
-        Ok(Self(z))
+        Ok(Self(size, z))
     }
 
     /// Get access to this memory.
-    pub fn lock(&mut self) -> LockedArrayGuard<'_, N> {
+    pub fn lock(&mut self) -> LockedArrayGuard<'_> {
         LockedArrayGuard::new(self)
     }
 }
 
+impl From<Vec<u8>> for LockedArray {
+    fn from(s: Vec<u8>) -> Self {
+        let mut out = Self::new(s.len()).unwrap();
+        out.lock().copy_from_slice(s.as_slice());
+        out
+    }
+}
+
 /// Locked memory that is unlocked for access.
-pub struct LockedArrayGuard<'g, const N: usize>(&'g mut LockedArray<N>);
+pub struct LockedArrayGuard<'g>(&'g mut LockedArray);
 
-impl<'g, const N: usize> LockedArrayGuard<'g, N> {
-    fn new(l: &'g mut LockedArray<N>) -> Self {
+impl<'g> LockedArrayGuard<'g> {
+    fn new(l: &'g mut LockedArray) -> Self {
         unsafe {
-            libsodium_sys::sodium_mprotect_readwrite(l.0);
+            libsodium_sys::sodium_mprotect_readwrite(l.1);
         }
         Self(l)
     }
 }
 
-impl<const N: usize> Drop for LockedArrayGuard<'_, N> {
+impl Drop for LockedArrayGuard<'_> {
     fn drop(&mut self) {
         unsafe {
-            libsodium_sys::sodium_mprotect_noaccess(self.0 .0);
+            libsodium_sys::sodium_mprotect_noaccess(self.0 .1);
         }
     }
 }
 
-impl<const N: usize> std::ops::Deref for LockedArrayGuard<'_, N> {
+impl std::ops::Deref for LockedArrayGuard<'_> {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { std::slice::from_raw_parts(self.0 .1 as *const u8, self.0 .0) }
+    }
+}
+
+impl std::ops::DerefMut for LockedArrayGuard<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe {
+            std::slice::from_raw_parts_mut(self.0 .1 as *mut u8, self.0 .0)
+        }
+    }
+}
+
+/// Locked memory with a size known at compile time.
+pub struct SizedLockedArray<const N: usize>(LockedArray);
+
+impl<const N: usize> SizedLockedArray<N> {
+    /// Create a new locked memory buffer.
+    pub fn new() -> Result<Self> {
+        Ok(Self(LockedArray::new(N)?))
+    }
+
+    /// Get access to this memory.
+    pub fn lock(&mut self) -> LockedArrayGuardSized<'_, N> {
+        LockedArrayGuardSized::new(self)
+    }
+}
+
+/// Locked memory that is unlocked for access.
+pub struct LockedArrayGuardSized<'g, const N: usize>(LockedArrayGuard<'g>);
```

Comment:
![sodoken_structs](https://github.com/user-attachments/assets/aba54ea7-3c89-4431-80d3-d36a286ce003)

```suggestion
/// Locked memory that is unlocked for access.
pub struct SizedLockedArrayGuard<'g, const N: usize>(LockedArrayGuard<'g>);
```

Reply:
Oh dear, yes fixed